# Day 1 - Python Basics

## Contents

- **Part 1 - Data Structures (45 min)**
    - 1.0. Introduction
    - 1.1. Data Types and Structures
    - 1.2. Lists: basis and manipulation
    - 1.3. Dictionaries: basis and manipulation
- **Test Part 1 (15 min)**
- **Part 2 - Compound Statements(60 min)**
    - 2.1. Conditionals (if, else, elif)
    - 2.2. Loops (for, while)
    - 2.3. Functions
- **Test Part 2 (15 min)**


# 1) Data structures




## 1.0) Introduction

### First steps
Writing code is an important part of programming. We start with a blank file, write a few lines of code, and a program is born! 


### Notebooks

They consist of units called `cells`. Cells can contain markdown text like this one, or Python code (notice that code cells are numbered). To run a cell you can click the `Run` button. These are some useful shortcuts:

- To run a cell and stay in the same cell, type `Ctrl+Enter` 
- To run a cell and move to the next cell, type `Shift+Enter`
- To run a cell and insert a new cell below, type `Alt+Enter`
- To run all the cells of a notebook, choose the `Cell -> Run all` menu
- Use the up and down arrow keys to move quickly between cells

**Remember**: **IDEs** on the other hand are software for building applications with the different tools for writing a program such as source code editor and code debugger. They facilitate this process with features like syntax highlighting and autocomplete. For further information chek this [note](https://keith-monreal.medium.com/notebook-vs-ide-180d38a0cd55#:~:text=Notebooks%20are%20files%20which%20supports,code%20editor%20and%20code%20debugger.)



To print or display any value or information in Python, you use the preset function **print**.

In [ ]:
print("Somos FICT, Somos ESPOL, bienvenidos a este Taller sobre Python")

In [ ]:
# Write the first code line that everyone has learnt: print("Hola mundo") 


In [ ]:
# Create a variable and name it saludo, assing the following to the variable "Hola mundo"

# Print the variable named saludo
    print()  #indentation

**Indentation**
Indentation refers to the spaces at the beginning of a code line.
Where in other programming languages the indentation in code is for readability only, the indentation in Python is very important.

**Comments**
Comments starts with a #, and Python will ignore them.


- Comments can be used to explain Python code.

- Comments can be used to make the code more readable.

- Comments can be used to prevent execution when testing code.

**Variables**
Variables are containers for storing data values.
Important: Variable names are case-sensitive.

## 1.1) Data types and structures

### Data types
In Python, we have different **types of data**, such as: float (floating point), int (integer), str (string) y bool (Boolean). 

- **float** - used for real numbers.
- **int** - used for integers.
- **str** - Text strings are defined in Python using single quotation marks
<code> 'Texto' </code> , comillas dobles <code> "Texto" </code> , o comillas triples <code> """Texto""" </code>

Strings generated by triple quotes can contain multiple lines of text, the newlines will be included in the value of the variable. That is why they are used to write the documentation of functions.
-  **bool** - A Boolean variable is a variable that can only take two possible values: True (verdadero) o False (falso). They are used to filter information from a data set.


In [ ]:
# The following are some examples of the main variables to be used in this workshop
current_year = 2023
average_height = 170.50
topic = "Fundamentos de Python"
succesful_workshop = True

It is possible to recognize the types (or classes) of variables using: **type(variable_name)**
This allows us to determine the type of variable.

In [ ]:
# print the types of the variable current_year

# print the types of all the variables with one print


### Math operations

Python can be used as a simple calculator:

In [ ]:
#Try some math operations


However, this only works for the basic arithmetic operations: addition ( + ), subtraction ( - ), multiplication ( \* ), division ( / ), exponentiation ( \*\* ), floor division ( // ), and modulo division ( % ). Notice that division (/) returns a float value, even if the two numbers are integers. Floor (or integer) division returns the quotient of the division, while modulo division returns the remainder of the division (though things get more complicated for negative numbers, see this [note](https://blog.teclado.com/pythons-modulo-operator-and-floor-division/)).

For more advanced operations, we need to import the `math` module:

The  **data structures** are arrays used to store several values, they can be assigned to variables, they are iterables, some are mutables and among the most common are:
- Lists
- Dictionaries
- Tuples
- Datasets
- Dataframes
- Numpy arrays
- Series

## 1.2) Lists

Lists are data structures that can store numbers, strings, lists, etc.
They are delimited by brackets ([a,b,c]) and their elements are separated by commas.

Some examples using **lists** are shown below for OPEC COUNTRIES 2023 (https://worldpopulationreview.com/country-rankings/opec-countries)

![Picture title](Recursos/OPEC_image.png)

*Figure 1. OPEC Logo*

In [ ]:
# List with 16 countries that have been members of OPEC
paises = ['Arabia Saudita', 'Venezuela', 'Kuwait', 'Iran', 'Irak', 'Libia', 'UAE', 'Algeria', 'Nigeria', 'Gabón', 'Ángola', 'Congo', 'Guinea Ecuatorial','Ecuador', 'Qatar', 'Indonesia']

# Other oil producing contries non OPEC members
NO_OPEP= ['Arzebaijan', 'Bahrein', 'Brunei', 'Kazajistán', 'Malasia', 'México', 'Omán', 'Rusia', 'Sudán', 'Sudán del Sur']

#retired/lapsed members
paises_retirados = ['Ecuador', 'Indonesia', 'Qatar']

## Python Operators

![Picture title](Recursos/Python_operators.png)
*Figure 2. Python Operators*

Lists can be manipulated, using arithmetic operators and methods.

**Keep in mind!** It is worth to mention that the methods are functions belonging to the object containing a list, because in Python every variable is related to an object, and likewise, every object belongs to a class. Therefore, methods are functions that are part of the **List** class.

### 1.2.1. Lists Indexing and Slicing

In [ ]:
#subsetting lists
# write a code to print the name of Libia using the list paises


#print(paises[5,6]) # indices must be integers or slices, not tuple
#print(paises[-11,-10]) # indices must be integers or slices, not tuple

#slicing lists [start:end] [inclusive: exclusive]
#slice the first 13 elements of the list paises


# Add list elements
# create a new list OPEP_plus adding + NO_OPEP to OPEP list


# Show the created lists


"Formatted string literals" f-strings are the string literals in which we use an f at the beginning and curly braces {} to enclose the expressions that will be replaced with other values.

In [ ]:
# Use f-strings to create a variable s,  s= #number of countries that have been members of OPEC
s = f'{} países han estado en la OPEP' #use len to provide the value

# Show s
print(s)

# Alternative way
print(f'')

Practice completing the code below:

In [ ]:
print(f'{} países se han retirado de la OPEP') #complete the code
print(f'{} países conforman la OPEP actualmente') #complete the code
print(f'{} países conforman la OPEP Plus') #complete the code

### 1.2.2) List Methods
Among the most important list methods, we have the following:

*append*, *remove*, *pop*, *del*, *clear*, *reverse*, *sort*, *index*, *extend*, *count*, *insert*

To apply a method you will use the following: list_name.method_name

![Picture title](Recursos/List_methods.png)
*Figure 3. List Methods*
**Fuente:**
https://www.w3schools.com/python/python_lists_methods.asp

In [ ]:
# The remove method removes the item with the specified value 
# Syntax > list.remove(element)
paises.('') #complete the code for removing Ecuador from the list paises
print(paises)

.remove('') #complete the code for removing Indonesia from the list paises
print(paises)

# Remove Qatar from the list paises

print(paises)

print("Estos son los paises que aun forman parte de la OPEP:")
print(paises)

In [ ]:
# The append method adds an element at the end of the list
# Syntax > list.append(element)

# show the list paises
print(paises)
# Include Ecuador in the list paises and show the updated list

print(paises)
print(f"{paises[-1]} ha sido el único país de latinoamérica que ha abandonado la OPEP por 2 veces")

In [ ]:
# Use f-strings and an operator to complete the code
print(f"Hay {str(len(paises) - 1)} paises que conforman actualmente la OPEP")

In [ ]:
# The pop() method removes the element at the specified position.
# Syntax> list.pop(position)

# write a code to print the number of countries in paises
print(len(paises))
# use pop to remove the name of the country in the last position of the list

print(len(paises))

In [ ]:
# The del keyword is used to delete objects (variables, lists, or parts of a list, etc.)
# Syntax> del list[pos]

# print the list of the Asian countries retired/lapsed from OPEC
print(paises_retirados)
# delete Ecuador from the list

print("Estos son los paises de Asia que decidieron retirarse de la OPEP:")
print(paises_retirados)

In [ ]:
# The reverse method reverses the sorting order of the elements
# Syntax. list.reverse()

# show the list paises
print(paises)
# reverse the order of the countries in the list paises

print("Paises de la OPEP ordenados desde el ultimo al primero")
print(paises)

In [ ]:
# The sort method sorts the list ascending by default. 
# Syntax > list.sort() 

# print the list paises with its elements alphabetically listed

print("Paises de la OPEP ordenados de forma alfabética:")
print(paises)

In [ ]:
# The extend method adds a list or elements (or any iterable) to the end of the current list
# Syntax. list.extend(iterable)

# show paises
print(paises)

# Create a list containing Brasil, Noruega and USA to the list paises
otros_paises = ('Brasil', 'Noruega', 'USA')

print(paises)

print("Estos no forman parte de la OPEP a pesar de ser paises con importante producción:")
print(paises[13:])

In [ ]:
# The insert() method inserts the specified value at the specified position
# Syntax > list.insert(position, element)
print(paises_retirados)

# add Ecuador to the list paises_retirados using insert

# show the list paises_retirados
print(paises_retirados)

In [ ]:
# The count() method returns the number of elements with the specified value
# Syntax > list.count(value)

# show the list paises_retirados
print(paises_retirados)

#print how many times apears Ecuador in the list

print(n_ecuador)

In [ ]:
# The clear() method removes all the elements from a list
# Syntax > list.clear()

# empty the list paises_retirados


print(paises_retirados)

## 1.4) Dictionaries: basis and manipulation

### Dictionaries
A dictionary is a data structure and a data type in Python with special characteristics that allows us to store any type of value such as integers, strings, lists and even other functions.
These dictionaries also allow us to identify each element by a key. To define a dictionary, the list of values is enclosed in braces. The key and value pairs are separated by commas, and the key and value are separated by colons.


**Volve Field**

In the southern Norwegian North Sea, about 200 km west of the city of Stavanger (see Figure 1), lies the Volve oil field, located in Block 15/9, at a depth of about 80 m. This field was developed by Statoil, now Equinor, with a 59.6% interest, while other project partners were ExxonMobil E&P Norway, with a 30.40% interest, and Bayerngas Norge, with a 10% interest. 

![volve](Recursos/volve.png)
*Figure 4. Volve Oilfield*


In [91]:
campo_Volve = {'País' : 'Noruega', 'Inicio_produccion': '2008',
               'Fin_produccion': '2016', 'Bloque' : '15/9','Tipo' : 'off-shore',
               'Pozo' :['F-1C','F-4','F-5','F-11','F-12','F-14', 'F-15 D']}
campo_Volve

{'País': 'Noruega',
 'Inicio_produccion': '2008',
 'Fin_produccion': '2016',
 'Bloque': '15/9',
 'Tipo': 'off-shore',
 'Pozo': ['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D']}

### Dictionaries manipulation
Among the most important dictionaries methods, we have the following:

*keys*, *values*, *items*, *get*, *update*, *pop*, *popitem*, *setdefault*

In [34]:
# The keys() method returns a view object
# The view object contains the keys of the dictionary, as a list
# The view object will reflect any changes done to the dictionary, see example below
print(campo_Volve.keys())
# Show the values of campo_Volve
print(campo_Volve.values())

dict_keys(['País', 'Inicio_produccion', 'Fin_produccion', 'Bloque', 'Tipo', 'Pozo'])
dict_values(['Noruega', '2008', '2016', '15/9', 'off-shore', ['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D']])


In [35]:
# The items() method returns a view object
# The view object contains the key-value pairs of the dictionary, as tuples in a list

# Show dictionary items
campo_Volve.items()

dict_items([('País', 'Noruega'), ('Inicio_produccion', '2008'), ('Fin_produccion', '2016'), ('Bloque', '15/9'), ('Tipo', 'off-shore'), ('Pozo', ['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D'])])

In [36]:
# The values() method returns a view object. 
#The view object contains the values of the dictionary, as a list.
# Show dictionary values
campo_Volve

{'País': 'Noruega',
 'Inicio_produccion': '2008',
 'Fin_produccion': '2016',
 'Bloque': '15/9',
 'Tipo': 'off-shore',
 'Pozo': ['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D']}

In [37]:
# The get method returns the value of the item with the specified key. 
# Syntax. dictionary.get(keyname, value)
print(campo_Volve)

# Show the value of the key 'País'
campo_Volve.get("País")

{'País': 'Noruega', 'Inicio_produccion': '2008', 'Fin_produccion': '2016', 'Bloque': '15/9', 'Tipo': 'off-shore', 'Pozo': ['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D']}


'Noruega'

In [38]:
# The update method inserts the specified items to the dictionary 
# Syntax > dictionary.update(iterable)
print(campo_Volve)

# Add the key: continente and assign the value: Europa and display it
campo_Volve.update({"continente":"Europa"})
campo_Volve

{'País': 'Noruega', 'Inicio_produccion': '2008', 'Fin_produccion': '2016', 'Bloque': '15/9', 'Tipo': 'off-shore', 'Pozo': ['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D']}


{'País': 'Noruega',
 'Inicio_produccion': '2008',
 'Fin_produccion': '2016',
 'Bloque': '15/9',
 'Tipo': 'off-shore',
 'Pozo': ['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D'],
 'continente': 'Europa'}

In [39]:
# The popitem method removes the item that was last inserted into the dictionary
# Syntax > dictionary.popitem()
print(campo_Volve)

# Remove the key: continente and its value: Europa
campo_Volve.popitem()
campo_Volve

{'País': 'Noruega', 'Inicio_produccion': '2008', 'Fin_produccion': '2016', 'Bloque': '15/9', 'Tipo': 'off-shore', 'Pozo': ['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D'], 'continente': 'Europa'}


{'País': 'Noruega',
 'Inicio_produccion': '2008',
 'Fin_produccion': '2016',
 'Bloque': '15/9',
 'Tipo': 'off-shore',
 'Pozo': ['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D']}

In [42]:
# The pop method removes the element at the specified position. Syntax > list.pop(pos)
print(campo_Volve)

# Add the key: continente and assign the value: Europa
campo_Volve['Continente'] = 'Europa'
campo_Volve
# Remove the key: continente and its value: Europa using only pop
campo_Volve.pop("Continente")
campo_Volve

{'País': 'Noruega', 'Inicio_produccion': '2008', 'Fin_produccion': '2016', 'Bloque': '15/9', 'Tipo': 'off-shore', 'Pozo': ['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D'], 'Continente': 'Europa'}


{'País': 'Noruega',
 'Inicio_produccion': '2008',
 'Fin_produccion': '2016',
 'Bloque': '15/9',
 'Tipo': 'off-shore',
 'Pozo': ['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D']}

In [45]:
campo_Volve.pop("Inicio_produccion")
campo_Volve

{'País': 'Noruega',
 'Fin_produccion': '2016',
 'Bloque': '15/9',
 'Tipo': 'off-shore',
 'Pozo': ['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D']}

In [49]:
# The setdefault method returns the value of a key (if the key is in dictionary). 
# If not, it inserts key with a value to the dictionary. 
# Syntax dictionary.setdefault(keyname, value)

campo_Volve.setdefault("Continente", "Europa")
# Use setdefault use result and show the dictionary
campo_Volve
# Use pop item to delete Europa 

{'País': 'Noruega',
 'Fin_produccion': '2016',
 'Bloque': '15/9',
 'Tipo': 'off-shore',
 'Pozo': ['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D'],
 'Inicio_produccion': '2008',
 'Continente': 'Europa'}

In [50]:
campo_Volve

{'País': 'Noruega',
 'Fin_produccion': '2016',
 'Bloque': '15/9',
 'Tipo': 'off-shore',
 'Pozo': ['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D'],
 'Inicio_produccion': '2008',
 'Continente': 'Europa'}

# Part 1 Test

#### 1) In this excercise you will create a list containing the names of the  academic programs offered by FICT ( and postgraduate programs) following the next steps:

**a)** Create a list called **pregrado** with the undergraduate programs and another liste named **postgrado** with the names of our postgraduate programs.


**b)** Concatenate and sort this list in alphabetical order

**c)** What is the name of the fifth program?

In [ ]:
# Insert your code



#### 2) Create a dictionary containing the average oil production in barrels per day of the following countries:
- United States -> 14,837,639.510
- China -> 4,905,070.874
- Kazakhstan -> 1,698,280.454
- Ecuador -> 548,378.765
- Cameroon -> 92,469.000
- Norway -> 2,003,747.525
- Brazil -> 3,242,957.836

Then, solve the following questions:

**a)** Add Qatar's oil production -> 1,987,192.978 bpd

**b)** Delete China's oil production from dictionary

**c)** What are the keys of this dictionary?

**d)** What is the length of this dictionary?

In [54]:
# Insert your code

country_oil = {"United Sates": "14,837,639.510", "China":"4,905,070.874", "Kazakhstan":"1,698,280.454",
              "Ecuador": "548,378.765", "Cameroon": "92,469.000", "Norway": "2,003,747.525",
              "Brazil": "3,242,957.836"}
country_oil

{'United Sates': '14,837,639.510',
 'China': '4,905,070.874',
 'Kazakhstan': '1,698,280.454',
 'Ecuador': '548,378.765',
 'Cameroon': '92,469.000',
 'Norway': '2,003,747.525',
 'Brazil': '3,242,957.836'}

In [56]:
# a)

country_oil["Qatar"] = "1,987,192.978"
country_oil

{'United Sates': '14,837,639.510',
 'China': '4,905,070.874',
 'Kazakhstan': '1,698,280.454',
 'Ecuador': '548,378.765',
 'Cameroon': '92,469.000',
 'Norway': '2,003,747.525',
 'Brazil': '3,242,957.836',
 'Qatar': '1,987,192.978'}

In [57]:
# b)

country_oil.pop("China")
country_oil

{'United Sates': '14,837,639.510',
 'Kazakhstan': '1,698,280.454',
 'Ecuador': '548,378.765',
 'Cameroon': '92,469.000',
 'Norway': '2,003,747.525',
 'Brazil': '3,242,957.836',
 'Qatar': '1,987,192.978'}

In [58]:
# c)

country_oil.keys()

dict_keys(['United Sates', 'Kazakhstan', 'Ecuador', 'Cameroon', 'Norway', 'Brazil', 'Qatar'])

In [59]:
# d)

len(country_oil)

7

# 2) Part 2 - Compound Statements

### 2.1) Conditionals (if, else, elif)

- The **if-else statement** is used to code the same way you would use it in the English language. The syntax for the if-else statement is:

- On the other hand, **The elif statement** allows you to check multiple expressions for TRUE and execute a block of code as soon as one of the conditions evaluates to TRUE.

- Similar to the else, **the elif statement** is optional. However, unlike else, for which there can be at most one statement, there can be an arbitrary number of **elif statements following an if.**

- The syntax of this compound statement is shown below:


![Picture Title](Recursos/if_sintaxis.png) 

In [61]:
# Using if and else example
campo = campo_Volve.copy()
campo['Nombre'] = 'Volve'

if campo["Nombre"] == "Volve": # insert the condition campo['Nombre'] = 'Volve'
    print("El campo Volve se encuentra en el area que corresponde al mar del norte Noruego")
else:
    print("No se de dispone de información del campo en mención")

El campo Volve se encuentra en el area que corresponde al mar del norte Noruego


In [64]:
campo_Volve

{'País': 'Noruega',
 'Fin_produccion': '2016',
 'Bloque': '15/9',
 'Tipo': 'off-shore',
 'Pozo': ['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D'],
 'Inicio_produccion': '2008',
 'Continente': 'Europa'}

In [65]:
#if #: # insert condition País = Inglaterra
if campo_Volve['Bloque'] == '15/9': 
    print('Este campo es Offshore')
elif campo_Volve['Bloque'] == '15/10':
    print('Este campo no existe')
else:
    print('No hay información sobre este país')

Este campo es Offshore


In [66]:
# Insert a list of API values 

API = [9, 15, 23, 32, 40]

# Identify wich type of crude is the first element of the API list

if API[0] <= 10:
    print("EL crudo es extra pesado")
elif API[0] >= 11 and API[0] <= 21:
    print("El crudo es pesado")
elif API[0] >= 22 and API[0] <= 29:
    print("El crudo es mediano")
elif API[0] >= 30:
    print("El crudo es liviano")

EL crudo es extra pesado


In [69]:
# # Identify wich type of crude is the element whose value is 23

if API[2] <= 10:
    print(f"EL crudo es extra pesado con un valor de {API[2]}")
elif API[2] >= 11 and API[2] <= 21:
    print(f"El crudo es pesado con un valor de {API[2]}")
elif API[2] >= 22 and API[2] <= 29:
    print(f"El crudo es mediano con un valor de {API[2]}")
elif API[2] >= 30:
    print(f"El crudo es liviano con un valor de {API[2]}")

El crudo es mediano con un valor de 23


### 2.2) Loops (for, while)

Loops are important in Python or in any other programming language as they help you to execute a **block of code repeatedly**. You will often come face to face with situations where you would need to use a piece of code over and over but you don't want to write the same line of code multiple times.

#### 2.2.1. for Loop

- A for loop is a programming concept that, when it's implemented, executes a piece of code over and over again "for" a certain number of times, based on a sequence. This "number of times" is determined by a sequence or an ordered list of things. 

- The syntax of this compound statement is shown below:


![Picture Title](Recursos/for_sintaxis.png) 

In [70]:
API

[9, 15, 23, 32, 40]

In [78]:
# Use a for Loop to determine the type of crude of the elements in the list API

for crude in API:
    if crude <= 10:
        print(f"El crudo es extrapesado {crude}")
    elif crude >= 11 and crude <= 21:
        print(f"El crudo es pesado {crude}")
    elif crude >= 22 and crude <= 29:
        print(f"El crudo es medino {crude}")
    elif crude >= 30:
        print(f"El crudo es liviano {crude}")

El crudo es extrapesado 9
El crudo es pesado 15
El crudo es medino 23
El crudo es liviano 32
El crudo es liviano 40


In [79]:
campo

{'País': 'Noruega',
 'Fin_produccion': '2016',
 'Bloque': '15/9',
 'Tipo': 'off-shore',
 'Pozo': ['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D'],
 'Inicio_produccion': '2008',
 'Continente': 'Europa',
 'Nombre': 'Volve'}

In [80]:
# Use for Loop to print the keys

for clave in campo.keys():
    print(clave)

País
Fin_produccion
Bloque
Tipo
Pozo
Inicio_produccion
Continente
Nombre


In [84]:
# Use for Loop to print the keys and values line by line
print(campo_Volve)
print()

for clave, valor in campo_Volve.items():
    print(f"{clave} : {valor}")

{'País': 'Noruega', 'Fin_produccion': '2016', 'Bloque': '15/9', 'Tipo': 'off-shore', 'Pozo': ['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D'], 'Inicio_produccion': '2008', 'Continente': 'Europa'}

País : Noruega
Fin_produccion : 2016
Bloque : 15/9
Tipo : off-shore
Pozo : ['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D']
Inicio_produccion : 2008
Continente : Europa


In [98]:
for value in campo_Volve.values():
    print(value)

Noruega
2008
2016
15/9
off-shore
['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D']


In [101]:
# Check the information of initial oil production date  

for value in campo_Volve.values():
    if value == 'Noruega':
        continue
    elif value != '2008':
        print('El campo empezó a producir a partir del 2008')
        break
    else:
        print('la Información del campo es correcta')

la Información del campo es correcta
El campo empezó a producir a partir del 2008


#### 2.2.2. while loop

A while loop is a programming concept that, when it's implemented, executes a piece of code over and over again while a given condition still holds true.

The above definition also highlights the three components that you need to construct the while loop in Python:

- The while keyword;
- A condition that transates to either True or False; And
- A block of code that you want to execute repeatedly


The syntax of this compound statement is shown below:

![Picture Title](Recursos/while_sintaxis.png) 

In [102]:
campo_Volve

{'País': 'Noruega',
 'Inicio_produccion': '2008',
 'Fin_produccion': '2016',
 'Bloque': '15/9',
 'Tipo': 'off-shore',
 'Pozo': ['F-1C', 'F-4', 'F-5', 'F-11', 'F-12', 'F-14', 'F-15 D']}

In [149]:
# use while loop to count the number of wells
count = 0

while count < len(campo_Volve["Pozo"]):
    count = count + 1
print(f"El campo volve tiene {count} pozos")

El campo volve tiene 7 pozos


### 2.3) Functions

A function is a group of related statements that performs a specific task.

On the other hand, functions help break our program into smaller and modular chunks. As our program grows larger and larger, functions make it more organized and manageable.

The syntax of this compound statement is shown below:

![Picture Title](Recursos/function_sintaxis.png) 

In [140]:
# Create a function to print 'es un campo Offshore que se encuentra en el Mar del Norte'
def campo_volve():
    print('es un campo Offshore que se encuentra en el Mar del Norte')

In [141]:
campo_volve()

es un campo Offshore que se encuentra en el Mar del Norte


In [144]:
# Use the previous function to complete the sentence
print(f'Uno de los campos mas conocidos del continente europeo es el campo volve el cual {campo_volve()}')

es un campo Offshore que se encuentra en el Mar del Norte
Uno de los campos mas conocidos del continente europeo es el campo volve el cual None


![darcy](Recursos/DarcyG.png)
*Figure 5. Darcy Law*

**Ecuación de Darcy**

if is horizontal linear flow:
$$Q = \frac{kA(P_{1}-P_{2})}{u L}$$  $$P_{1}>P_{2}$$

if is inclined linear flow:
$$Q = \frac{-kA}{u} ( \frac{dp}{dL} + {g\rho sin \theta} )$$

**Where:** \
*$Q$:* Flow rate $(m^{3}/s)$\
*$k$:* Coefficient of permeability $(m^{2})$\
*$A$:* Cross sectional area of flow $(m^{2})$\
*$P1$:* Pressure at point 1 $(Pa)$\
*$P2$:* Pressure at point 1 $(Pa)$\
*$u$:* Viscosity  $(Pa *s)$\
*$L$:* Length $(m)$

In [126]:
# Import math library
import math

# Linear flow rate

def flow_rate(k, A, P1, P2, u, L, g, pho=None, theta=None):
    if pho is None and theta is None:
        q = ((k * A) * (P1 - P2)) / (u * L) # Flujo horizontal
        
    elif pho is not None and theta is not None:
        q = ((-k * A) / u) * (((P1 - P2) / L) + g * pho * math.sin(theta)) # Flujo inclinado
        
    return q

**Water Resistivity (Archie equation)**

$$R_w = \frac{R_{t} * \phi^{m}}a$$

**Where:** \
*$R_w$:* Apparent Resistivity of formation water $(\Omega-m)$\
*$R_t$:* True formation resistivity as derived from a deep reading resistivity log $(\Omega-m)$\
*$\phi$:* Porosity $(fraction)$\
*$a$:* Tortuosity factor\
*$m$:* cementation exponent

In [127]:
# Water resistivity - Archie equation
def resistividad_agua(ro, phi, a, m):
    rw = (ro * phi ** m) / a
    return rw

# Part 2 Test

#### 1) What is the flow rate in m$^3$/s through a core if we measured the following values under horizontal flow test:
- $K = 2*{10^{-11}}{m^2}$
- $P_{1} = 18000 Pa$
- $P_{2} = 10000 Pa$
- $A = 0.25 {m^2}$
- $u= 0.001 mPa*s$
- $L= 0.5 m$
1 darcy is aproximately 1×10−12 $m^{2}$

In [131]:
# Insert your code
Q = flow_rate(2E-11, 0.25, 18000, 10000, 0.001, 0.5, 9.8)

print(f"Flow rate is: {Q} m3/s")

Flow rate is: 8e-05 m3/s



#### 2) Write a code to obtain the artificial lift methods whose name starts with a vowel. For this example, consider the following list: ["Rod pump", "Gas lift", "PCP", "ESP", "Hydraulic pump"]

In [ ]:
# Insert your code

# 3) Bonus data type
## 3.1) datetime
The Volve field start producing in January of 2008 to June 2016. Let's check for how many days the field produced!

In [132]:
import datetime
start_production = datetime.datetime(year=2008, month=1, day=1)
end_production = datetime.datetime(year=2016, month=6, day=1)
prod_days = end_production - start_production
print(prod_days)

3074 days, 0:00:00


Wow! It produced for 3074 days! What if the field started production today? Until what year would the production last?

In [133]:
now = datetime.date.today()
print(now + prod_days)

2031-09-19
